### ISSUE TITLE
How does redux-form binds the redux actions with form fields dynamically?

### ISSUE THREAD

**Issue by antoaravinth** [erikras/redux-form#293]
Thanks for a great plugin for html forms with redux! This is not an issue, but want to understand how things are working in your code. 

So without your plugin, this is how we can tie up with any html form element with redux data store:

```
export default class Form extends Component{

    constructor(props) {
        super(props)
    }
    render() {

        const { data } = this.props
        const { dispatch } = this.props;

        return (
            <div>
                <input type="textbox" onChange={(event) =>  dispatch(type(event.target.value))}  />
            </div>
        )
    }
}

function mapStateToProps(state) {
    const { data } = state;

    return {
        data
    }
}

export default connect(mapStateToProps)(Form)
```

And in your `reducers.js`:

```
function formRedux(state = {}, action) {
    switch (action.type){
        case "FORMDATA":
            //do something here
        default:
            return Object.assign({} , state , {
                data : ""
            });
    }
}

const rootReducer = combineReducers({
  . . . 
  formRedux
})
```

And in your action you have:

```
function typeSomething(data){
    return {
        type: "FORMDATA",
        value: data
    }
}

export function type(data) {
    return (dispatch, getState) => {
        return dispatch(typeSomething(data))
    }
}
```

And once this is ready, render your component and type something, your data store is going to be updated. 

But when using your plugin we do like this:

```
ContactForm = reduxForm({ // <----- THIS IS THE IMPORTANT PART!
  form: 'contact',                           // a unique name for this form
  fields: ['firstName', 'lastName', 'email'] // all the fields in your form
})(ContactForm);
```

I want to understand, how your dynamically attaching the redux actions on the fly using the `fields` name? 

Thanks for the response.

**Comment by erikras** [comment]
Magic! :tophat: :rabbit2: :smile:

I'm not 100% sure what you mean by "dynamically attaching the redux actions on the fly". The `this.props.fields.firstName.onChange` function fires off a `{ type: 'CHANGE', form: 'contact', field: 'firstName' }` action. The "attaching" to the input occurs by you, the users of `redux-form` adding `onChange={this.props.fields.firstName.onChange}` to the input manually. The `fields` array  you give to `reduxForm()` is used to generate the `this.props.fields` object given to your wrapped component.

If I totally missed what you meant, please rephrase the question more specifically.

**Comment by antoaravinth** [comment]
@erikras : Lol! Yeah sorry, now I got it; I didn't see the code properly. Last question, now imagine, I want to set the redux state only on `blur` event on a input element, how I can say this to `redux-form`?

**Comment by erikras** [comment]
Easy. Rather than the default:
```js
<input type="text" {...name}/>
```
You will need to manually supply the values that you want, thus omitting the `onChange`:
```js
<input type="text" value={name.value} onBlur={name.onBlur}/>
```
`value` and `onBlur` are the only two that you _really_ need, but there is also `name` and `onFocus` that you could pass through.

Alternatively, you could override `onChange` only by placing it after the destructured prop:
```js
<input type="text" {...name} onChange={() => {}}/>
```
Make sense?

**Comment by antoaravinth** [comment]
@erikras : Thats brilliant. Thanks for your response !

**Comment by styrmis** [comment]
I'm also experiencing this kind of slowness with a form, where each keystroke in a text input is causing re-rendering to occur. Multiple keystrokes then result in the UI hanging for a second as it catches up. This occurs with and without production settings / dev tools enabled.

The solution of not binding the `onChange` handler has helped, but clearly something else in my setup is causing this slowness—unfortunately I'm not yet sure what that is.

One thing, I needed to specify `<input type="text" defaultValue={name.value} onBlur={name.onBlur}/>`, rather than specify `value`, as if I did that then the input field wouldn't respond to any input, as the value wouldn't be changing.

**Comment by jcalfee** [comment]
Would be nice to have a config boolean that would remove the `onChange` and `value` ..

**Comment by erikras** [comment]
Nice article, @DanielMSchmidt. 👍

**Comment by JeroenNelen** [comment]
I also switched to using defaultValue and onBlur, it makes a significant different in performance for me as well.
I only noticed that this doesn't play well together with the Material-UI typings, as it doesn't expect a defaultValue and an onBlur

**Comment by jsj14** [comment]
Strangely formActions become undefined when using deep forms. Is there a way to bind it to the child form component so that it remains accessible?

**Comment by quytang** [comment]
I have a question that why redux form using action to dispatch change? Should it use callback instead? It's a strange to me and don't feel comfortable with it.
